docs/sql-migration-guide-upgrade.md

```diff
@@ -132,6 +132,10 @@ license: |
 
   - Since Spark 3.0, Spark will cast `String` to `Date/TimeStamp` in binary comparisons with dates/timestamps. The previous behaviour of casting `Date/Timestamp` to `String` can be restored by setting `spark.sql.legacy.typeCoercion.datetimeToString` to `true`.
 
+  - Since Spark 3.0, when Avro files are written with user provided schema, the fields will be matched by field names between catalyst schema and avro schema instead of positions.
+
+  - Since Spark 3.0, when Avro files are written with user provided non-nullable schema, even the catalyst schema is nullable, Spark is still able to write the files. However, Spark will throw runtime NPE if any of the records contains null.
```

Comment:
Throwing a runtime exception by default looks risky to me. Could we change back the original behavior? We can add a conf if the end users can accept the runtime exception for nullability constraints.

Reply:
cc @cloud-fan @gengliangwang @HyukjinKwon

Reply:
What was the original behaviour?

Reply:
Let us add a legacy conf and throw an exception by default.

Reply:
cc @xuanyuanking could you submit a follow-up PR?

Reply:
Sure

Reply:
What's the legacy behavior in Spark 2.4? @gengliangwang

Reply:
@cloud-fan before this PR, there should be an error in the `resolveNullableType` if user specifies a non-nullable Avro schema to write a nullable dataframe.
After this PR, there will be warning message and possible runtime error.

Reply:
i see, basically we move query compile-time error to runtime error.

Reply:
After further investigation, I think the 3.0 behavior here is good enough, cause this behavior only take effect while no records contain null. Also, while any records contain null, the new approach will give a better exception.

Let's use the UT added in this PR to illustrate:
For the old behavior, we'll get exception `org.apache.avro.AvroRuntimeException: Not a union: "int"` which didn't express the problem of null data insertion.
For the current approach, the error message will be `NullPointerException: in test_schema in string null of string in field Name of test_schema`.

So if both legacy and non-legacy mode throw exception, and the legacy mode message is less clear than the new one, the legacy config might not necessary. WDYT?